src/Course/State.hs

```diff
@@ -135,44 +122,137 @@ instance Monad (State s) where
 --
 -- >>> let p x = (\s -> (const $ pure (x == 'i')) =<< put (1+s)) =<< get in runState (findM p $ listh ['a'..'h']) 0
 -- (Empty,8)
-findM ::
-  Monad f =>
-  (a -> f Bool)
-  -> List a
-  -> f (Optional a)
-findM =
-  error "todo: Course.State#findM"
+--
+--
+findM :: Monad f => (a -> f Bool) -> List a -> f (Optional a)
+findM _ Nil = return Empty
+findM p (x :. xs) =
+  p x
+  >>= \b -> if b then return (Full x) else findM p xs
+
+-- attempt 2: foldRight, but maybe this forces it to inspect every value in the list rather than short circuit?
+-- findM :: Monad f => (a -> f Bool) -> List a -> f (Optional a)
+-- findM p xs = foldRight folder (return Empty) xs
+--   where
+--     folder x result =
+--       p x >>= \b -> if b then return (Full x) else result
+
+-- rewriting this...
+--
+-- let p x = (\s -> (const $ pure (x == 'c')) =<< put (1+s)) =<< get in runState (findM p $ listh ['a'..'h']) 0
+-- let p x = get >>= (\s -> (const $ pure (x == 'c')) =<< put (1+s)) in runState (findM p $ listh ['a'..'h']) 0
+-- let p x = get >>= \s -> put (1 + s) >>= \_ -> pure (x == 'c') in runState (findM p (listh ['a'..'h'])) 0
+--                    0         1 + 0       ()         False
+--                    1         1 + 1       ()         False
+--                    2         1 + 2       ()         True
 
 -- | Find the first element in a `List` that repeats.
 -- It is possible that no element repeats, hence an `Optional` result.
---
+
 -- /Tip:/ Use `findM` and `State` with a @Data.Set#Set@.
 --
 -- prop> \xs -> case firstRepeat xs of Empty -> let xs' = hlist xs in nub xs' == xs'; Full x -> length (filter (== x) xs) > 1
 -- prop> \xs -> case firstRepeat xs of Empty -> True; Full x -> let (l, (rx :. rs)) = span (/= x) xs in let (l2, r2) = span (/= x) rs in let l3 = hlist (l ++ (rx :. Nil) ++ l2) in nub l3 == l3
-firstRepeat ::
-  Ord a =>
-  List a
-  -> Optional a
-firstRepeat =
-  error "todo: Course.State#firstRepeat"
+--
+-- attempt 1: my misconception of what was being asked because I couldn't
+-- read the test cases
```

Comment:
Yeah, the instructions are pretty annoyingly unclear sometimes aren't they? It took me about 20 minutes to try to figure out exactly what the last one meant.